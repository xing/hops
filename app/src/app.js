import React from 'react';
import Match from 'react-router/Match';


import { headline } from './styles.css';


const Home = () => (
  <h1 className={ headline }>Hello World!</h1>
);

const App = () => (
  <div>
    <Match exactly pattern="/" component={ Home } />
  </div>
);


export default App;
