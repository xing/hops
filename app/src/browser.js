import React from 'react';
import { render } from 'react-dom';
import BrowserRouter from 'react-router/BrowserRouter'

import App from './app';


export default () => {
  render(
    <BrowserRouter><App/></BrowserRouter>,
    document.querySelector('#main')
  );
};
