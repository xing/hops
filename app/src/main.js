
import React, { Component, PropTypes } from 'react';
import { Route } from 'react-router';
import { connect } from 'react-redux';

import { createAction, createReducer, render } from 'hops';

import { headline } from './style.css';

const namespace = 'home';
@connect(
  state => state[namespace],
  {
    update: createAction(namespace)
  }
)
class Home extends Component {
  static propTypes = {
    greeting: PropTypes.string,
    update: PropTypes.func
  };
  componentDidMount() {
    const { update } = this.props;
    update({'greeting': {'$set': 'Hello World!'}});
  }
  render() {
    const { greeting } = this.props;
    return (
      <h1 className={ headline }>{ greeting }</h1>
    );
  }
}

const routes = (
  <Route path='/' component={ Home } />
);
const reducers = {[namespace]: createReducer(namespace)};

export default render({ routes, reducers });
