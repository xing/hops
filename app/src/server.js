import React from 'react';
import { renderToString } from 'react-dom/server';
import { ServerRouter, createServerRenderContext } from 'react-router';


import App from './app';

import template from './template.tpl';

import { version } from '../package.json';


export default (req, res) => {

  const context = createServerRenderContext()

  let markup = renderToString(
    <ServerRouter location={ req.url } context={ context }>
      <App/>
    </ServerRouter>
  );

  const result = context.getResult();

  if (result.redirect) {

    res.writeHead(301, { Location: result.redirect.pathname });

    res.end();
  }
  else {
    if (result.missed) {

      res.writeHead(404);

      markup = renderToString(
        <ServerRouter location={ req.url } context={ context }>
          <App/>
        </ServerRouter>
      );
    }

    res.write(template({ markup, version }));
    res.end();
  }
};
