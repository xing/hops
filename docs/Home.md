Welcome to the hops wiki!

Contents

* [Getting Started Guide](./getting-started)
* [Concepts](./concepts)
* [Packages](./modules)

