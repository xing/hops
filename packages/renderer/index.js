'use strict';

var events = require('events');

var express = require('express');
var mocks = require('node-mocks-http');

var hopsConfig = require('hops-config');
var createMiddleware = require('hops-middleware');

module.exports = function createRenderer(webpackConfig, watchOptions) {
  if (arguments.length === 1 && 'webpackConfig' in webpackConfig) {
    webpackConfig = webpackConfig.webpackConfig;
    watchOptions = webpackConfig.watchOptions;
    console.warn(
      'Calling createRenderer() with just an options hash is deprecated.',
      'Please refer to the latest docs:',
      'https://github.com/xing/hops/blob/master/packages/renderer/README.md'
    );
  }
  var router = new express.Router();
  if (hopsConfig.bootstrapServer) {
    hopsConfig.bootstrapServer(router, hopsConfig);
  }
  router.use(createMiddleware(webpackConfig, watchOptions));
  return function(options) {
    if (typeof options === 'string') {
      options = { url: options };
    }
    return new Promise(function(resolve, reject) {
      var req = mocks.createRequest(options);
      var res = mocks.createResponse({
        eventEmitter: events.EventEmitter,
        request: req,
      });
      res.on('finish', function() {
        if (res.statusCode !== 200) {
          reject(new Error('invalid status code: ' + res.statusCode));
        } else {
          resolve(res._getData());
        }
      });
      router.handle(req, res, function(error) {
        if (error) {
          reject(error);
        } else {
          resolve();
        }
      });
    });
  };
};
