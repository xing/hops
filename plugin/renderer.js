/**
 * @file plugin/renderer
 *
 * @author Daniel Dembach <daniel@dmbch.net>
 */
'use strict';

var path = require('path');

var webpack = require('webpack');
var MemoryFS = require('memory-fs');

/** @ignore */
exports.eval = function (fileContent, filePath, callback) {
  try {
    Object.keys(require.cache).forEach(function(key) {
      delete require.cache[key];
    });
    var context = Object.assign(
      new module.constructor(),
      { paths: module.paths }
    );
    // eslint-disable-next-line no-underscore-dangle
    context._compile(fileContent.toString(), filePath);
    callback(null, context.exports.default || context.exports);
  }
  catch (error) {
    callback(error);
  }
};

/** @ignore */
exports.createRenderer = function (configFile) {
  if (!configFile) {
    return Promise.resolve(function () {
      return Promise.resolve({ state: '{}', dom: ''});
    });
  }
  return Promise.resolve().then(function () {
    return new Promise(function (resolve, reject) {
      var mfs = new MemoryFS();
      var config = require(configFile);
      var compiler = webpack(config.clean ? config.clean() : config);
      compiler.outputFileSystem = mfs;
      compiler.run(function(compileError) {
        if (compileError) { reject(compileError); }
        else {
          var filePath = path.join(config.output.path, config.output.filename);
          mfs.readFile(filePath, function (readFileError, fileContent) {
            if (readFileError) { reject(readFileError); }
            else {
              exports.eval(fileContent, filePath, function (evalError, result) {
                if (evalError) { reject(evalError); }
                else { resolve(result); }
              });
            }
          });
        }
      });
    });
  });
};
