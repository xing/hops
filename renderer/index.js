'use strict';

var events = require('events');

var mocks = require('node-mocks-http');

var createMiddleware = require('../middleware');

module.exports = function createRenderer (hopsConfig, watchOptions) {
  var middleware = createMiddleware(hopsConfig, watchOptions);

  return function (location) {
    return new Promise(function (resolve, reject) {
      var req = mocks.createRequest({
        url: location
      });
      var res = mocks.createResponse({
        eventEmitter: events.EventEmitter,
        request: req
      });
      res.on('finish', function () {
        if (res.statusCode !== 200) {
          reject('invalid status code: ' + res.statusCode);
        } else {
          resolve(res._getData());
        }
      });
      middleware(req, res, function (error) {
        if (error) { reject(error); } else { resolve(); }
      });
    });
  };
};
