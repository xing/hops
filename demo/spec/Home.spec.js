import React from 'react';
import renderer from 'react-test-renderer';
import { Home } from '../src/Home';

it('renders correctly', () => {
  const tree = renderer.create(<Home message="ey yo captain jack" />).toJSON();
  expect(tree).toMatchSnapshot();
});
