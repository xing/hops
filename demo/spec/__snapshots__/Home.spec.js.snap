// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <h1
    className="headline"
  >
    ey yo captain jack
  </h1>
</div>
`;
